File: RL-Glue/Network/Agent/RL_client_agent.py
import sys
import time
from RL_agent import *
from RL_netlib import *

def onAgentInit(sock):
	taskSpec = sock.recvString()
	agent_init(taskSpec)

def onAgentStart(sock):
	theObservation = sock.recvADT()
	theAction = agent_start(theObservation)
	sock.sendADT(theAction)

def onAgentStep(sock):
	ro = sock.recvRewardObservation()
	theAction = agent_step(ro.r, ro.o)
	sock.sendADT(theAction)

def onAgentEnd(sock):
	reward = sock.recvDouble()
	agent_end(reward)

def onAgentCleanup(sock):
	agent_cleanup()

def onAgentFreeze(sock):
	agent_freeze()

def onAgentMessage(sock):
	inMessage = sock.recvString()
	outMessage = agent_message(inMessage)
	sock.sendString(outMessage)

def runAgentEventLoop(sock):
	agentState = 0
	while agentState != kAgentCleanup:
		agentState = sock.recvInt()
		if agentState == kAgentInit:
			onAgentInit(sock)
		elif agentState == kAgentStart:
			onAgentStart(sock)
		elif agentState == kAgentStep:
			onAgentStep(sock)
		elif agentState == kAgentEnd:
			onAgentEnd(sock)
		elif agentState == kAgentCleanup:
			onAgentCleanup(sock)
		elif agentState == kAgentFreeze:
			onAgentFreeze(sock)
		elif agentState == kAgentMessage:
			onAgentMessage(sock)
		else:
			sys.stderr.write(kUnknownMessage % (agentState))

isDaemon = 0
for arg in sys.argv:
	if arg == "--stayalive":
		isDaemon = True
sock = waitForConnection(kLocalHost,kDefaultPort,kRetryTimeout)
sock.sendInt(kAgentConnection)
runAgentEventLoop(sock)
sock.close()
while isDaemon:
	sock = waitForConnection(kLocalHost,kDefaultPort,kRetryTimeout)
	sock.sendInt(kAgentConnection)
	runAgentEventLoop(sock)
	sock.close()

File: RL-Glue/Network/Agent/RL_client_agent.java
public class RL_client_agent 
{
    protected static final String kUnknownMessage = "Unknown Message: ";
    protected RL_netlib netlib;
    protected AgentType agent;

    public RL_client_agent() 
    {
	agent = new AgentType();
    }

    protected void onAgentInit() 
    {
	String taskSpec = netlib.getInputStream().readString();
	agent.agent_init(taskSpec);
    }

    protected void onAgentStart() 
    {
	final RLAbstractType observation = netlib.readAbstractType();
	final RLAbstractType action = agent.agent_start(observation);
	netlib.writeAbstractType(action);
    }

    protected void onAgentStep()
    {
	final double reward = netlib.readDouble();
	final RLAbstractType observation = netlib.readAbstractType();
	final RLAbstractType action = agent.agent_step(reward, observation);
	netlib.writeAbstractType(action);
    }

    protected void onAgentEnd()
    {
	final double reward = netlib.readDouble();
	agent.agent_end();
    }

    protected void onAgentCleanup()
    {
	agent.agent_cleanup();
    }

    protected void onAgentFreeze() 
    {
	agent.agent_freeze();
    }

    protected void onAgentMessage()
    {
	final String recvMessage = netlib.readString();
	final String sendMessage = agent.agent_message(recvMessage);

	/* this is not finished */
    }

    /* Should fix this and the RL_netlib constructor to throw more specific exceptions */
    protected void connect(String host, int port, int timeout) throws Exception
    {	
	netlib = new RL_netlib(host, port, timeout);
    }

    protected void runAgentEventLoop(Socket theSocket) 
    {
	int agentState = 0;

	InputDataStream input = theSocket.getInputDataStream();
	agentState = input.readInt();

	do {

	    switch(agentState) {
	    case kAgentInit:
		onAgentInit();
		break;
		
	    case kAgentStart:
		onAgentStart();
		break;
		
	    case kAgentStep:
		onAgentStep();
		break;
		
	    case kAgentEnd:
		onAgentEnd();
		break;
		
	    case kAgentCleanup:
		onAgentCleanup();
		break;
		
	    case kAgentFreeze:
		onAgentFreeze();
		break;
		
	    case kAgentMessage:
		onAgentMessage();
		break;
		
	    default:
		System.err.println(kUnknownMessage + agentState + "\n");
		System.exit(1);
		break;
	    };

	} while (agentState != kAgentCleanup);
    }

    protected void close() 
    {
	netlib.close();
    }

    public static void main(String [] args) 
    {
	RL_client_agent client = new RL_client_agent();
	boolean autoReconnect = false;

	do {
	    client.connect(host, port, kRetryTimeout);
	    client.runAgentEventLoop();
	    client.close();
	} while (autoReconnect);
    }
}